### PR TITLE
[Paddle] Add deterministic option in DotProductAttention

### PR DESCRIPTION
**Pull request by Wong4j**
# Description

The custormers need an option to enable deterministic Attention. 
When user sets ~~FLAGS_cudnn_deterministic=1~~  `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, workspace optimization is enabled.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/paddle/layer/attention.py [hunk 1]
```diff
@@ -567,6 +575,22 @@ def _te_forward(
         set_zero: bool = True,
     ) -> paddle.Tensor:
 
+        if self.deterministic:
+            # workspace optimization path is deterministic
+            os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "-1"
```
Comment (jeng1220) [inline transformer_engine/paddle/layer/attention.py:580]: Should it be `os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "1"` ?

**Comment by cyanguwa** [comment]
Hi @Wong4j , could you use the same environment variable name as in PyTorch please? `NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT `. Just so we are more consistent across TE. 

https://github.com/NVIDIA/TransformerEngine/blob/6ee92c4bcfdc78bd0c3c29bbbaf3c02d1fcacd51/transformer_engine/pytorch/attention.py#L4053-L4067

Thanks.

**Comment by Wong4j** [comment]
@cyanguwa The environment variable names are already consistent. I just copied this part from pytorch.

File: transformer_engine/paddle/layer/attention.py (hunk 1 above)
Comment (Wong4j) [reply transformer_engine/paddle/layer/attention.py:580]: `-1` means `enable workspace optimization always`. This part is copied directly from pytorch.

**Comment by jeng1220** [comment]
@Wong4j will implement `NVTE_ALLOW_NONDETERMINISTIC_ALGO` in this PR.

**Comment by jeng1220** [comment]
LGTM

**Comment by jeng1220** [comment]
LGTM

**Comment by jeng1220** [comment]
@cyanguwa ,
Could you please merge the code if anything looks fine?

**Comment by cyanguwa** [comment]
@Wong4j , could you take a look at the failed CI please? Thanks!

File: transformer_engine/paddle/layer/attention.py [hunk 2]
```diff
@@ -458,6 +464,24 @@ def __init__(
...
+        if "NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT" in os.environ:
+            if os.environ["NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT"] == "0":
+                os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "0"
+            if os.environ["NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT"] == "1":
+                os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "-1"
+
+        if self.deterministic:
+            # workspace optimization path is deterministic
+            os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "-1"
+
```
Comment (cyanguwa) [inline transformer_engine/paddle/layer/attention.py:484]: Maybe add a note here about the cuDNN versions? Something like this? https://github.com/NVIDIA/TransformerEngine/blob/8e039fdcd98fc56582d81e373880c1509c2b8f73/transformer_engine/pytorch/attention.py#L5148C1-L5150C72

Also, could you please pass the `self.deterministic` value down to `nvte_fused_attn_bwd_xxx` calls?  Thanks.

Comment (Wong4j) [reply transformer_engine/paddle/layer/attention.py:498]: Done. BTW, I changed the `if (cudnn_runtime_version >= 90000 && sm_arch_ >= 90)` to `if (cudnn_runtime_version >= 90000)` here[fused_attn_f16_arbitrary_seqlen.cu](https://github.com/NVIDIA/TransformerEngine/pull/956/files#diff-28710e6e65e3526663810821d4131eaeccb3e3807262a8fbd9ecfa5799739cfb)
. `set_deterministic_algorithm` is also applicable and essential for Ada and Ampere.

**Pull request merged**